notifier/sendloop.go

```diff
@@ -0,0 +1,245 @@
+// Copyright The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notifier
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+type sendLoop struct {
+	alertmanagerURL string
+
+	cfg    *config.AlertmanagerConfig
+	client *http.Client
+	opts   *Options
+
+	metrics *alertMetrics
+
+	mtx      sync.RWMutex
+	queue    []*Alert
+	hasWork  chan struct{}
+	stopped  chan struct{}
+	stopOnce sync.Once
+
+	logger *slog.Logger
+}
+
+func newSendLoop(alertmanagerURL string, client *http.Client, cfg *config.AlertmanagerConfig, opts *Options, logger *slog.Logger, metrics *alertMetrics) *sendLoop {
+	return &sendLoop{
+		alertmanagerURL: alertmanagerURL,
+		client:          client,
+		cfg:             cfg,
+		opts:            opts,
+		logger:          logger,
+		metrics:         metrics,
+		queue:           make([]*Alert, 0, opts.QueueCapacity),
+		hasWork:         make(chan struct{}, 1),
+		stopped:         make(chan struct{}),
+	}
+}
+
+func (s *sendLoop) add(alerts ...*Alert) {
+	select {
+	case <-s.stopped:
+		// TODO(queueperam): maybe we should increment dropped?
```

Comment:
I think we can remove this; `...due to errors when sending to Alertmanager` from "dropped" description.